common/sgx/verifier.c

```diff
@@ -302,98 +339,78 @@ static oe_result_t _fill_with_known_claims(
-    size_t claims_index = 0;
-
-    if (claims_length < header->num_claims)
-        OE_RAISE(OE_INVALID_PARAMETER);
-
-    claims_buf += sizeof(*header);
-    claims_buf_size -= sizeof(*header);
-    for (uint64_t i = 0; i < header->num_claims; i++)
-    {
-        oe_sgx_plugin_claims_entry_t* entry =
-            (oe_sgx_plugin_claims_entry_t*)claims_buf;
-        uint64_t size;
-
-        // Sanity check sizes.
-        if (claims_buf_size < sizeof(*entry))
-            OE_RAISE(OE_CONSTRAINT_FAILED);
-
-        OE_CHECK(oe_safe_add_u64(sizeof(*entry), entry->name_size, &size));
-        OE_CHECK(oe_safe_add_u64(size, entry->value_size, &size));
-
-        if (claims_buf_size < size)
-            OE_RAISE(OE_CONSTRAINT_FAILED);
-
-        // Finally, add the claim.
-        OE_CHECK(_add_claim(
-            &claims[claims_index++],
-            entry->name,
-            entry->name_size,
-            entry->name + entry->name_size,
-            entry->value_size));
-
-        // Go to next entry.
-        claims_buf += size;
-        claims_buf_size -= size;
-    }
+    // Default hash for empty string, as described in the literature.
+    static uint8_t sha256_for_empty_string[] = {
+        0xe3, 0xb0, 0xc4, 0x42, 0x98, 0xfc, 0x1c, 0x14, 0x9a, 0xfb, 0xf4,
+        0xc8, 0x99, 0x6f, 0xb9, 0x24, 0x27, 0xae, 0x41, 0xe4, 0x64, 0x9b,
+        0x93, 0x4c, 0xa4, 0x95, 0x99, 0x1b, 0x78, 0x52, 0xb8, 0x55};
+
+    // Produce a hash of the custom_claims.
+    // If there is no data, set to default value
+    if (!custom_claims || !custom_claims_size)
+        memcpy(hash_out, sha256_for_empty_string, sizeof(*hash_out));
+    else
+        // Produce a hash of the custom_claims.
+        OE_CHECK(oe_sha256(custom_claims, custom_claims_size, hash_out));
 
     result = OE_OK;
 
 done:
-    if (result != OE_OK)
-    {
-        for (size_t i = 0; i < claims_index; i++)
-            _free_claim(&claims[i]);
-    }
     return result;
 }
 
 oe_result_t oe_sgx_extract_claims(
+    const sgx_evidence_format_type_t format_type,
     const oe_uuid_t* format_id,
-    const uint8_t* evidence,
-    size_t evidence_size,
+    const uint8_t* report_body,
+    size_t report_body_size,
+    const uint8_t* custom_claims,
+    size_t custom_claims_size,
     const oe_sgx_endorsements_t* sgx_endorsements,
     oe_claim_t** claims_out,
     size_t* claims_length_out)
 {
     oe_result_t result = OE_UNEXPECTED;
-    oe_report_header_t* header = (oe_report_header_t*)evidence;
-    oe_sgx_plugin_claims_header_t* claims_header = NULL;
-    size_t report_size = sizeof(*header) + header->report_size;
     oe_claim_t* claims = NULL;
     uint64_t claims_length = 0;
     uint64_t claims_size = 0;
     size_t claims_added = 0;
+    size_t additional_claim = 0;
+    sgx_report_data_t report_data;
 
-    // Check if the buffer is the proper size.
-    if (evidence_size - report_size < sizeof(*claims_header))
+    if (!format_id || !report_body || !report_body_size)
```

Comment:
Consider adding the buffer and size consistency check that was applied to `endorsements` elsewhere: `(!custom_claims == !custom_claims_size)`

Reply:
Thanks!

Reply:
That's an interesting choice, why do we allow that? There's no functional value in doing so, and providing a non-zero size to a null pointer to an array should still be trapped.

Reply:
Since custom claims are attached at the end of the SGX quote, some callers can just get a pointer to right after the quote for the custom claims and calculate the size of the custom claims (which could be zero). In this case the pointer is not NULL, but the size is zero.

Yes, let me add the trap to null pointer with non-zero size.

Reply:
Thanks for the explanation, that would be useful to add as a comment here, since new contributors or maintainers may block this behavior in the future without understanding the context.